libraries/AP_Math/scurves.cpp

```diff
@@ -0,0 +1,645 @@
+    } else {
+        return 0.0f;
+    }
+}
+
+// time left before sequence will complete
+float scurves::get_time_remaining() const
+{
+    if (segtype == SegmentType::STRAIGHT) {
+        return get_time_remaining_straight();
+    } else if (segtype == SegmentType::SPLINE) {
+        return get_time_remaining_spline();
+    } else {
+        return 0.0f;
+    }
+}
+
+// return true if the sequence is braking to a stop
+bool scurves::braking() const
+{
+    if (segtype == SegmentType::STRAIGHT) {
+        return braking_straight();
+    } else if (segtype == SegmentType::SPLINE) {
+        return braking_spline();
+    } else {
+        return true;
+    }
+}
+
+// Straight implementations
+
+// increment time pointer and return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_pva_straight(float dt, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_straight(_t, time_scale, pos, vel, accel);
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_to_pva_straight(float dt, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_straight(_t, time_scale, pos, vel, accel);
+    pos -= _track;
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_time_pva_straight(float time, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    float scurve_P1, scurve_V1, scurve_A1, scurve_J1;
+    bool finish = update(time, scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_1 * scurve_P1;
+    vel += _delta_unit_1 * scurve_V1 * MIN(time_scale * 1.1f, 1.0f);
+    accel += _delta_unit_1 * scurve_A1 * time_scale;
+    return finish;
+}
+
+// Spline implementations
+
+// increment time pointer and return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_pva_spline(float dt, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_spline(_t, time_scale, pos, vel, accel);
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the destination
+bool scurves::move_to_pva_spline(float dt, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_spline(_t, time_scale, pos, vel, accel);
+    pos -= _track;
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_time_pva_spline(float time, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    bool finish;
+    float scurve_P1, scurve_V1, scurve_A1, scurve_J1;
```

Comment:
Move these to where they're being assigned to for the first time.

Reply:
these are now declared directly above where they are used.